### DISCUSSION TITLE
Simple question about custom instruction implementation

### DISCUSSION THREAD

**Discussion by Rhythmicc** [gem5/gem5#921] [Q&A]
I want to add instructions in `src/arch/riscv/isa/decoder.isa`. My software will save the array's address to the register and then access the array in a similar way:
```isa
0x1: VectorNonSplitFormat::vmv_x_s({{
    Rd_ud = 0;
    uint16_t*meta = (uint16_t*) Vs2_vu, a = meta[0];
    uint8_t nnzA = __builtin_popcount(a);
    printf("bitmap: 0x%04x, nnzA: %d\n", a, nnzA);
    fflush(stdout);

    double *da = (double *)(Vs2_vu[1]), *db = (double *)(Vs2_vu[2]), *dc = (double *)(Vs2_vu[3]);
    printf("av: %p\n", da);
    printf("bv: %p\n", db);
    printf("cv: %p\n", dc);
    fflush(stdout);

    for (uint8_t i = 0; i < nnzA; ++i) printf("%lf ", da[i]);
    printf("\n");
    fflush(stdout);

    Rd_ud = 1;
}}, OPMVV, VectorMiscOp);
```
But when executing, I encountered a segmentation fault. I believe this is a problem with the way I'm accessing memory. Could you please advise on how to correctly access the values in the array?

**Comment by ivanaamit** [comment]
@Rhythmicc I recommend checking out this tutorial, as it might provide helpful insights and answers to your question: [gem5 bootcamp 2022: Instruction execution and adding instructions](https://www.youtube.com/watch?v=Z5B02jkNpck)

**Reply by Rhythmicc** [reply]
Thank you very much, I double checked this video; I can successfully read data from the address stored in the vector register using `xc->readMem`, but I can't write the result of the calculation back to memory via `xc->writeMem`. In my tests, `xc->writeMem` doesn't seem to do anything and doesn't report any errors, as I showed it in a new problem: #923
